Fix settings panel reset during state update

System233 opened the pull request:
<!-- **Note:** Consider creating PRs as a DRAFT. For early feedback and self-review. -->

## Description

* Add Save/Done buttons to allow users to cancel setting changes.  
* Decouple ApiOptions from ExtensionStateContext, enabling the settings panel to edit on a snapshot of the apiConfiguration, fixing the issue of settings panel reset during API response completion, profile renaming, and any state updates.
* Improve the event handler for `handleInputChange` in ApiOptions.
* Only save settings when clicking the Save button to avoid accidental changes.
* Switching profiles will no longer save settings

![image](https://github.com/user-attachments/assets/bf314076-07bf-4ce9-b6d4-3a3b45b12b04)


## Type of change

<!-- Please ignore options that are not relevant -->

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

<!-- Please describe the tests that you ran to verify your changes -->

## Checklist:

<!-- Go over all the following points, and put an `x` in all the boxes that apply -->

- [x] My code follows the patterns of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

## Additional context

<!-- Add any other context or screenshots about the pull request here -->

## Related Issues

<!-- List any related issues here. Use the GitHub issue linking syntax: #issue-number -->

## Reviewers

<!-- @mention specific team members or individuals who should review this PR -->

<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Fixes settings panel reset issue by decoupling `ApiOptions` from `ExtensionStateContext` and adding Save/Done buttons for explicit configuration changes.
> 
>   - **Behavior**:
>     - Adds Save/Done buttons in `SettingsView.tsx` to allow users to save or discard changes.
>     - Decouples `ApiOptions` from `ExtensionStateContext`, allowing editing on a snapshot of `apiConfiguration`.
>     - Changes in `ApiOptions.tsx` ensure settings are only saved when Save is clicked.
>     - Switching profiles no longer saves settings automatically.
>   - **Functions**:
>     - Introduces `onApiConfigurationUpdate` in `SettingsView.tsx` to handle API configuration updates.
>     - Refactors `handleInputChange` in `ApiOptions.tsx` to use `onApiConfigurationUpdate`.
>   - **Tests**:
>     - Updates tests in `ApiOptions.test.tsx` and `SettingsView.test.tsx` to reflect new Save/Done behavior.
>     - Ensures tests check for the presence of Save button and correct message posting to VSCode.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=RooVetGit%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for a62e6a78fdd15b483f54befb7bcb6c20964b5389. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

System233 commented:
> Maybe it should warn you if you switch profiles or leave the settings tab with unsaved changes?

I am considering it, but webview-ui cannot call `vscode.window.showInformationMessage`, is there any good way?

cte commented:
> I am considering it, but webview-ui cannot call `vscode.window.showInformationMessage`, is there any good way?

We might want to use https://ui.shadcn.com/docs/components/alert-dialog

I can make it look at little bit nicer / less heavy-handed.

System233 commented:
> > I am considering it, but webview-ui cannot call , is there any good way?`vscode.window.showInformationMessage`
> 
> We might want to use https://ui.shadcn.com/docs/components/alert-dialog
> 
> I can make it look at little bit nicer / less heavy-handed.

simply adjusted the transparency and it looks good now.

![image](https://github.com/user-attachments/assets/f7431e8f-43a8-4646-a807-919038d63554)

cte commented:
> simply adjusted the transparency and it looks good now.

Looks nice!

System233 commented:
Due to the optimization of the save logic, `saveApiConfiguration` is no longer needed for now.

Final result: 
![image](https://github.com/user-attachments/assets/c5d2f98c-c9f7-46e3-90e0-7c714590ac48)

cte commented on webview-ui/src/components/ui/comfirm-dialog.tsx:
```diff
@@ -0,0 +1,58 @@
+	onResult: (confirm: boolean) => void
+	onClose: () => void
+}
+export const ConfirmDialog = ({ onResult, onClose, icon, show, title, message }: ConfirmDialogProps) => {
```
Not a blocker, but we might want to use the official shadcn version of this component:

```bash
npx shadcn@latest add alert-dialog
```

System233 replied:
Many `radix-ui` components are already in use, it also provides an `Alert Dialog`, and we only use the `Dropdown` from the `vscrui` , I’m considering switching these to `radix-ui`.

samhvw8 replied:
@System233  i think we can use shadcn too, shadcn built on radix-ui with tailwind theme support, and @cte  already config theme for tailwind and shadcn component

System233 replied:
👌

cte commented on webview-ui/src/components/settings/__tests__/SettingsView.test.tsx:
```diff
@@ -302,7 +301,7 @@ describe("SettingsView - Allowed Commands", () => {
 	})
 
-	it("saves allowed commands when clicking Done", () => {
+	it("saves allowed commands when clicking Save", () => {
```
I think this test is failing:

```
  ● SettingsView - Allowed Commands › saves allowed commands when clicking Save

    expect(jest.fn()).toHaveBeenCalled()

    Expected number of calls: >= 1
    Received number of calls:    0

      328 |                     }),
      329 |             )
    > 330 |             expect(onDone).toHaveBeenCalled()
          |                            ^
      331 |     })
      332 | })
      333 |

      at Object.<anonymous> (src/components/settings/__tests__/SettingsView.test.tsx:330:18)
```

cte commented on webview-ui/src/components/settings/ApiOptions.tsx:
```diff
@@ -61,6 +69,21 @@ const ApiOptions = ({ apiErrorMessage, modelIdErrorMessage, fromWelcomeView }: A
+	const inputEventHandler = <E,>(event: E) => (event as { target: HTMLInputElement })?.target?.value as any
+	const noTransformHandler = <T,>(value: T) => value
+	const dropdownEventHandler = <T,>(event: DropdownOption | string | undefined) =>
+		(typeof event == "string" ? event : event?.value) as T
```
Nit-pick: I'd probably name these something like:
- `inputEventTransform`
- `dropdownEventTransform`
- `noTransform`

I think this makes it clear that they are transform function rather than actual event handlers.

System233 replied:
OK

System233 commented:
Adjusted the following UI to make it compatible with all themes

![image](https://github.com/user-attachments/assets/71a990d8-789e-4917-a2fd-ddc722d20ba6)
![image](https://github.com/user-attachments/assets/9c52c555-8c6b-4ae5-bdec-7b9d4f8fb361)
![image](https://github.com/user-attachments/assets/b8d0ebbe-c8b2-458a-8b37-6636468bc439)
![image](https://github.com/user-attachments/assets/2b7ff59b-fbff-4bb8-9823-658eb544bea3)
![image](https://github.com/user-attachments/assets/83fb2788-8edf-41d4-93d6-9d0de596c922)
![image](https://github.com/user-attachments/assets/737623ca-97e0-40db-876a-e8cd848b57b8)
![image](https://github.com/user-attachments/assets/45032cff-50c1-4384-b569-5e981ee0b18c)

## Unit-test
I'm not sure what this is
![image](https://github.com/user-attachments/assets/2192411f-eb1c-4c34-862d-12c1d3484373)

I added mock data to api_conversation_history.json and then I got this
```
<--- Last few GCs --->

[36880:000002621C559000]    47972 ms: Scavenge 4056.6 (4125.2) -> 4045.0 (4128.0) MB, pooled: 0 MB, 5.77 / 0.00 ms  (average mu = 0.163, current mu = 0.194) allocation failure;
[36880:000002621C559000]    48861 ms: Mark-Compact 4059.8 (4128.5) -> 4046.9 (4131.7) MB, pooled: 0 MB, 876.10 / 0.00 ms  (average mu = 0.194, current mu = 0.224) allocation failure; scavenge might not succeed


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
----- Native stack trace -----

 1: 00007FF6058520AD node::SetCppgcReference+17245
 2: 00007FF6057BA4D8 v8::base::CPU::num_virtual_address_bits+92376
 3: 00007FF60633A1B1 v8::Isolate::ReportExternalAllocationLimitReached+65
 4: 00007FF606327096 v8::Function::Experimental_IsNopFunction+2790
 5: 00007FF606176920 v8::internal::StrongRootAllocatorBase::StrongRootAllocatorBase+31392
 6: 00007FF606170644 v8::internal::StrongRootAllocatorBase::StrongRootAllocatorBase+6084
 7: 00007FF60616BCF5 v8::CpuProfileNode::GetScriptResourceNameStr+188069
 8: 00007FF605AF13BD BIO_ssl_shutdown+189
```

System233 commented:
![image](https://github.com/user-attachments/assets/ef8573f4-fbfb-4766-91a8-9c7975e1d7a5)
![image](https://github.com/user-attachments/assets/16f0ef45-8a95-4906-afda-2a421fe67a33)

mrubens commented:
Tests are flaky/broken in main unfortunately

mrubens reviewed (approved):
👏

mrubens commented on webview-ui/src/components/settings/TemperatureControl.tsx:
```diff
@@ -48,27 +49,15 @@ export const TemperatureControl = ({ value, onChange, maxValue = 1 }: Temperatur
 						<input
-							aria-label="Temperature control text input"
-							type="text"
+							type="range"
```
Actually I'm not sure about changing this to a slider - not all of the providers have the same 0-2 range. And separately, I'm noticing an error where the value gets reset to 1.

https://github.com/user-attachments/assets/be3fcdcf-0f18-4a66-8488-e4492f76e13e

mrubens replied:
Can we just leave as a text input for now?

System233 replied:
sure. 1 is the default value, other values ​​can also be the default values

System233 replied:
range 0-2 comes from the parent component, and default value is set when isCustomTemperature is toggled, these behaviors are the same as before the change, is it really bad after changing to a slider?

System233 replied:
https://github.com/user-attachments/assets/ef727be2-2038-42b4-842e-20dbdc984c01

mrubens replied:
Ah, I guess I was confused since 0 used to be the default value. I don't feel super strongly against the slider and it's certainly nicer in some ways, I just think it implies a valid range of 0-2 that's not the case for all models.

System233 replied:
> Ah, I guess I was confused since 0 used to be the default value. I don't feel super strongly against the slider and it's certainly nicer in some ways, I just think it implies a valid range of 0-2 that's not the case for all models.

I'm not sure either, I didn't change maxValue, and I've restored the default value to 0

mrubens replied:
If you like the slider better let’s go for it then

mrubens commented:
Actually I'm seeing some weirdness on rename now as well
https://github.com/user-attachments/assets/cb8a757f-4595-4ab3-a72f-77901307e79d

System233 commented:
> 实际上，我现在也看到了一些关于 rename 的奇怪之处
> 
>  Screen.Recording.2025-02-19.at.9.12.18.PM.mov

It was fine before, I'll fix it

> Actually I'm seeing some weirdness on rename now as well
> 
>  Screen.Recording.2025-02-19.at.9.12.18.PM.mov

It was fine before, I'll fix it

mrubens commented:
Rename seems to be working better, thanks! Looks like just one unit test, assuming that failure is legit.

System233 commented:
> Rename seems to be working better, thanks! Looks like just one unit test, assuming that failure is legit.

The slider is debounced, so there need to delay and assert it.
My mistake, fix it now.

mrubens commented:
🚀

System233 commented:
@mrubens I found that there were still references to the ExtensionStateContext in the ModelPickers and those settings were not being snapshotted, working on fixing that.
